### PR TITLE
*: fix datum's collation behavior to build correct logical ranges

### LINKED ISSUE #14688: ranger: update `Range` to support collations in TiDB.
## Feature Request

**Describe the feature you'd like:**
In order to provide full collation support in TiDB, `Range`(and codes related to it) should be updated accordingly. For example, a range condition of strings should be converted to their mem-comparable bytes(the "sortKeys") calculated by the collations.

**Teachability, Documentation, Adoption, Migration Strategy:**
Please see TiDB RFC #14574 for more details on collations support in TiDB.

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Set the collation correctly so that the ranger can get correct collation information.

This PR closes #14688

### What is changed and how it works?

When create new string/enum/set datum, don't forget to set it's collation in production env.

### Check List 

Tests 

 - Unit test(Will be added later)
 - Manual test (The same with the upcoming unit tests)

Code changes

 - Has exported function/method change

### CODE REVIEW

File: sessionctx/context.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ type Context interface {
 	// GetClient gets a kv.Client.
 	GetClient() kv.Client
 
-	// SetValue saves a value associated with this context for key.
+	// SetValueForTest saves a value associated with this context for key.
```
Comment (wjhuang2016) [inline sessionctx/context.go]: ditto

File: table/column.go [hunk 2]
```diff
@@ -183,6 +183,7 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo) (
 		return casted, err
 	}
 
+	// TODO: we may need to handle non-binary collation.
```
Comment (wjhuang2016) [inline table/column.go]: We don't need to handle non-binary colltion here, they are handled in other places.

File: types/datum.go [hunk 3]
```diff
@@ -278,10 +278,12 @@ func (d *Datum) GetMysqlEnum() Enum {
 }
 
 // SetMysqlEnum sets Enum value
-func (d *Datum) SetMysqlEnum(b Enum) {
+func (d *Datum) SetMysqlEnum(b Enum, collation string, length int) {
 	d.k = KindMysqlEnum
 	d.i = int64(b.Value)
 	sink(b.Name)
+	d.collation = collation
+	d.length = uint32(length)
```
Comment (wjhuang2016) [inline types/datum.go]: This may cause bugs, please remove this line. And better to remove it from arguments.

File: types/datum.go [hunk 4]
```diff
@@ -292,10 +294,12 @@ func (d *Datum) GetMysqlSet() Set {
 }
 
 // SetMysqlSet sets Set value
-func (d *Datum) SetMysqlSet(b Set) {
+func (d *Datum) SetMysqlSet(b Set, collation string, length int) {
 	d.k = KindMysqlSet
 	d.i = int64(b.Value)
 	sink(b.Name)
+	d.collation = collation
+	d.length = uint32(length)
```
Comment (wjhuang2016) [inline types/datum.go]: ditto

File: types/datum.go [hunk 5]
```diff
@@ -420,8 +424,8 @@ func (d *Datum) GetValue() interface{} {
 	}
 }
 
-// SetValue sets any kind of value.
-func (d *Datum) SetValue(val interface{}) {
+// SetValueForTest sets any kind of value.
+func (d *Datum) SetValueForTest(val interface{}) {
```
Comment (wjhuang2016) [inline types/datum.go]: What's the difference between SetValue and SetValueForTest?

File: types/datum.go [hunk 6]
```diff
@@ -1672,9 +1730,9 @@ func invalidConv(d *Datum, tp byte) (Datum, error) {
 func NewDatum(in interface{}) (d Datum) {
 	switch x := in.(type) {
 	case []interface{}:
-		d.SetValue(MakeDatums(x...))
+		d.SetValueForTest(MakeDatums(x...))
```
Comment (wjhuang2016) [inline types/datum.go]: Why use ForTest here?

File: table/column.go (hunk 2 above)
Comment (winoros) [reply table/column.go]: Oh it's an outdated comment

File: types/datum.go (hunk 5 above)
Comment (winoros) [reply types/datum.go]: There're many tests called it without any collation information.

Comment (winoros) [reply types/datum.go]: There're many tests called it without any collation information.

File: sessionctx/context.go (hunk 1 above)
Comment (qw4990) [inline sessionctx/context.go]: The comment should be updated.

**Comment by bb7133** [comment]
Please address the comments.

File: util/ranger/points.go [hunk 7]
```diff
@@ -205,18 +208,41 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []point {
 		err   error
 		ft    *types.FieldType
 	)
+	_, collation, _ := expr.CharsetAndCollation(expr.GetCtx())
 	if col, ok := expr.GetArgs()[0].(*expression.Column); ok {
```
Comment (wjhuang2016) [inline util/ranger/points.go]: I  think we have better put these two branches into a function, their logics are quite similar.

File: util/collate/collate.go [hunk 8]
```diff
@@ -35,6 +35,17 @@ const (
 	DefaultLen = 0
 )
 
+// SameCollate checks whether the two collate are the same.
+func SameCollate(collate1, collate2 string) bool {
```
Comment (wjhuang2016) [inline util/collate/collate.go]: `Same` may cause confusion, please use other name like "compatible".

**Pull request merged**